### PR TITLE
reland: Started using FlutterEngineGroups by default on Android

### PR DESCRIPTION
**Pull request by gaaclarke**
relands: https://github.com/flutter/engine/pull/37822
revert pr: https://github.com/flutter/engine/pull/38351

The fix is in a separate commit in the PR.  I just factored out some logic for creating engines from groups from elsewhere that takes into account the entrypoint.  It's a bit tricky since it isn't clear how the engine was getting those things.

There already is a test that covers this.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by gaaclarke** [comment]
oops, prematurely uploaded, I'll fix it.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java [hunk 1]
```diff
@@ -220,6 +228,18 @@ void onAttach(@NonNull Context context) {
...
+  private FlutterEngineGroup.Options addEntrypointOptions(FlutterEngineGroup.Options options) {
+    String appBundlePathOverride = host.getAppBundlePath();
+    if (appBundlePathOverride == null || appBundlePathOverride.isEmpty()) {
+      appBundlePathOverride = FlutterInjector.instance().flutterLoader().findAppBundlePath();
+    }
+
+    DartExecutor.DartEntrypoint dartEntrypoint =
+        new DartExecutor.DartEntrypoint(
+            appBundlePathOverride, host.getDartEntrypointFunctionName());
+    return options.setDartEntrypoint(dartEntrypoint).setInitialRoute(host.getInitialRoute());
```
Comment (Nayuta403) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:240]: Here is missing `host.getDartEntrypointArgs()` :-)

**Review by Nayuta403 — COMMENTED** [review]
LGTM 👍🏻

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java [hunk 2]
```diff
@@ -109,8 +110,15 @@ public void onFlutterUiNoLongerDisplayed() {
...
   FlutterActivityAndFragmentDelegate(@NonNull Host host) {
+    this(host, null);
+  }
+
+  FlutterActivityAndFragmentDelegate(@NonNull Host host, @Nullable FlutterEngineGroup engineGroup) {
     this.host = host;
     this.isFirstFrameRendered = false;
+    if (engineGroup != null) {
+      this.engineGroup = engineGroup;
+    }
```
Comment (ColdPaleLight) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:121]: nit: Removing the if judgment will make the code simpler

```suggestion
    this.engineGroup = engineGroup;
```

**Review by ColdPaleLight — APPROVED** [review]
LGTM!

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:121]: done

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java]: It makes sense to me that that should be here too.  Good catch, added.

**Comment by chinmaygarde** [comment]
@gaaclarke Did you or someone else re-run the check that tripped on "Resource not accessible by integration". I'm chasing down the flake and just want to make sure it didn't automagically resolve itself.

**Comment by gaaclarke** [comment]
> @gaaclarke Did you or someone else re-run the check that tripped on "Resource not accessible by integration". I'm chasing down the flake and just want to make sure it didn't automagically resolve itself.

@chinmaygarde wasn't me.  I didn't even see it.  I pushed up a handful of commits back-to-back, so maybe you saw it in between the pushes?

**Comment by gaaclarke** [comment]
Oh wait, i see the message about autosubmit failing.  Weird, I don't know what happened.

**Comment by chinmaygarde** [comment]
Okie dokie. Pushing the commits was probably what caused the re-runs. Anyway, I have the reference.

**Pull request merged**